File: appLab4B/src/main/java/com/github/hummel/cga/lab4b/RenderPanel.java
package com.github.hummel.cga.lab4b;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

public class RenderPanel extends JPanel {
    private final List<Triangle> triangles;
    private final Matrix4 viewportMatrix;
    private final Matrix4 projectionMatrix;
    private final Camera camera;
    private final BufferedImage bufferedImage;
    private final double[] zBuffer;
    private final Graphics2D imgGraphics;
    private Matrix4 viewMatrix;
    private Matrix4 modelMatrix;
    private Matrix4 camMatrix;
    private int prevMouseX;
    private int prevMouseY;
    private double rotateY;
    private double rotateX;

    public RenderPanel(List<Triangle> triangles) {
        this.triangles = triangles;
        viewportMatrix = MatrixBuilder.buildViewport(Main.width, Main.height);
        projectionMatrix = MatrixBuilder.buildProjection(1.75, 90.0);
        double dist = 5.0;
        camera = new Camera();
        camera.eye = new Vector4(dist * Math.cos(rotateX) * Math.cos(rotateY), dist * Math.sin(rotateX), dist * Math.cos(rotateX) * Math.sin(rotateY));
        camera.target = new Vector4(0.0, 0.0, 0.0);
        camera.up = new Vector4(0.0, 1.0, 0.0);
        viewMatrix = MatrixBuilder.buildView(camera);
        bufferedImage = new BufferedImage(Main.width, Main.height, BufferedImage.TYPE_INT_RGB);
        imgGraphics = bufferedImage.createGraphics();
        zBuffer = new double[Main.width * Main.height];
        imgGraphics.setBackground(new Color(0, 0, 0, 0));

        setupCamMatrix();
        addMouseMotionListener(new MouseAdapter() {
            @Override
            public void mouseDragged(MouseEvent e) {
                super.mouseDragged(e);

                if (rotateX > Math.PI / 2 - 0.1) {
                    rotateX = Math.PI / 2 - 0.05;
                } else if (rotateX < -Math.PI / 2 + 0.1) {
                    rotateX = -Math.PI / 2 - 0.05;
                }
                rotateY -= (e.getX() - prevMouseX) / 200.0;
                rotateX -= (e.getY() - prevMouseY) / 200.0;

                prevMouseX = e.getX();
                prevMouseY = e.getY();
            }

            @Override
            public void mouseMoved(MouseEvent e) {
                super.mouseMoved(e);
                prevMouseX = e.getX();
                prevMouseY = e.getY();
            }
        });
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2d = (Graphics2D) g;
        g.setColor(Color.GREEN);

        double dist = Main.dist;
        camera.eye = new Vector4(dist * Math.cos(rotateX) * Math.cos(rotateY), dist * Math.sin(rotateX), dist * Math.cos(rotateX) * Math.sin(rotateY));
        viewMatrix = MatrixBuilder.buildView(camera);
        setupCamMatrix();

        Matrix4 finalMatrix = camMatrix;

        imgGraphics.clearRect(0, 0, Main.width, Main.height);
        var filteredList = AlgoUtils.filterTriangles(triangles, camera);
        var drawList = AlgoUtils.applyMatrix(filteredList, finalMatrix);
        for (Triangle triangle : drawList) {
            AtomicInteger i = new AtomicInteger(0);
            double[] depthArr = new double[3];
            for (Vector4 vector4 : triangle.vertices) {
                depthArr[i.getAndIncrement()] = vector4.get(3);
                vector4.divSelf(vector4.get(3));
            }
            triangle.depthArr = depthArr;
        }

        Collection<Triangle[]> newList = new ArrayList<>();
        for (int i = 0; i < filteredList.size(); i++) {
            newList.add(new Triangle[]{filteredList.get(i), drawList.get(i)});
        }
        Arrays.fill(zBuffer, Double.POSITIVE_INFINITY);

        newList.parallelStream().forEach(trianglePair -> {
            AlgoUtils.drawRasterTriangle(bufferedImage, trianglePair[0], trianglePair[1], zBuffer, camera);
        });

        g2d.drawImage(bufferedImage, 0, 0, Main.width, Main.height, null);
    }

    private void setupCamMatrix() {
        camMatrix = viewportMatrix.mul(projectionMatrix).mul(viewMatrix);
    }
}

File: appLab4B/src/main/kotlin/com/github/hummel/cga/lab4b/RenderPanel.kt
package com.github.hummel.cga.lab4b

import com.github.hummel.cga.lab4b.MatrixBuilder.buildProjection
import com.github.hummel.cga.lab4b.MatrixBuilder.buildView
import com.github.hummel.cga.lab4b.MatrixBuilder.buildViewport
import java.awt.Color
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.image.BufferedImage
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import javax.swing.JPanel
import kotlin.math.cos
import kotlin.math.sin

class RenderPanel(private val faces: List<Face?>?) : JPanel() {
	private val viewportMatrix = buildViewport(Main.width, Main.height)
	private val projectionMatrix = buildProjection(1.75, 90.0)
	private val camera: Camera
	private val bufferedImage: BufferedImage
	private val zBuffer: DoubleArray
	private val imgGraphics: Graphics2D
	private var viewMatrix: Matrix4
	private var camMatrix: Matrix4? = null
	private var prevMouseX = 0
	private var prevMouseY = 0
	private var rotateY = 0.0
	private var rotateX = 0.0

	init {
		val dist = 5.0
		camera = Camera()
		camera.eye =
			Vertex(dist * cos(rotateX) * cos(rotateY), dist * sin(rotateX), dist * cos(rotateX) * sin(rotateY))
		camera.target = Vertex(0.0, 0.0, 0.0)
		camera.up = Vertex(0.0, 1.0, 0.0)
		viewMatrix = buildView(camera)
		bufferedImage = BufferedImage(Main.width, Main.height, BufferedImage.TYPE_INT_RGB)
		imgGraphics = bufferedImage.createGraphics()
		zBuffer = DoubleArray(Main.width * Main.height)
		imgGraphics.background = Color(0, 0, 0, 0)

		setupCamMatrix()
		addMouseMotionListener(object : MouseAdapter() {
			override fun mouseDragged(e: MouseEvent) {
				super.mouseDragged(e)

				if (rotateX > Math.PI / 2 - 0.1) {
					rotateX = Math.PI / 2 - 0.05
				} else if (rotateX < -Math.PI / 2 + 0.1) {
					rotateX = -Math.PI / 2 - 0.05
				}
				rotateY -= (e.x - prevMouseX) / 200.0
				rotateX -= (e.y - prevMouseY) / 200.0

				prevMouseX = e.x
				prevMouseY = e.y
			}

			override fun mouseMoved(e: MouseEvent) {
				super.mouseMoved(e)
				prevMouseX = e.x
				prevMouseY = e.y
			}
		})
	}

	override fun paintComponent(g: Graphics) {
		super.paintComponent(g)
		val g2d = g as Graphics2D
		g.setColor(Color.GREEN)

		val dist = Main.dist
		camera.eye =
			Vertex(dist * cos(rotateX) * cos(rotateY), dist * sin(rotateX), dist * cos(rotateX) * sin(rotateY))
		viewMatrix = buildView(camera)
		setupCamMatrix()

		val finalMatrix = camMatrix

		imgGraphics.clearRect(0, 0, Main.width, Main.height)
		val filteredList = AlgoUtils.filterTriangles(faces, camera)
		val drawList = AlgoUtils.applyMatrix(filteredList, finalMatrix!!)
		for (face in drawList) {
			val i = AtomicInteger(0)
			val depthArr = DoubleArray(3)
			for (vertex in face.vertices) {
				depthArr[i.getAndIncrement()] = vertex[3]
				vertex.divSelf(vertex[3])
			}
			face.depthArr = depthArr
		}

		val newList: MutableCollection<Array<Face?>> = ArrayList()
		for (i in filteredList.indices) {
			newList.add(arrayOf(filteredList[i], drawList[i]))
		}
		Arrays.fill(zBuffer, Double.POSITIVE_INFINITY)

		newList.parallelStream().forEach { facePair: Array<Face?> ->
			AlgoUtils.drawRasterTriangle(
				bufferedImage,
				facePair[0]!!,
				facePair[1]!!,
				zBuffer,
				camera
			)
		}

		g2d.drawImage(bufferedImage, 0, 0, Main.width, Main.height, null)
	}

	private fun setupCamMatrix() {
		camMatrix = viewportMatrix.mul(projectionMatrix).mul(viewMatrix)
	}
}
